daemon/containerd/image_delete.go

```diff
@@ -53,123 +54,180 @@ import (
 //
 // TODO(thaJeztah): image delete should send prometheus counters; see https://github.com/moby/moby/issues/45268
 func (i *ImageService) ImageDelete(ctx context.Context, imageRef string, force, prune bool) ([]imagetypes.DeleteResponse, error) {
-	parsedRef, err := reference.ParseNormalizedNamed(imageRef)
-	if err != nil {
-		return nil, err
+	var c conflictType
+	if !force {
+		c |= conflictSoft
 	}
 
-	img, err := i.resolveImage(ctx, imageRef)
+	img, all, err := i.resolveAllReferences(ctx, imageRef)
 	if err != nil {
 		return nil, err
 	}
 
-	imgID := image.ID(img.Target.Digest)
-
-	explicitDanglingRef := strings.HasPrefix(imageRef, imageNameDanglingPrefix) && isDanglingImage(img)
-	if isImageIDPrefix(imgID.String(), imageRef) || explicitDanglingRef {
-		return i.deleteAll(ctx, img, force, prune)
-	}
-
-	singleRef, err := i.isSingleReference(ctx, img)
-	if err != nil {
-		return nil, err
-	}
-	if !singleRef {
-		err := i.client.ImageService().Delete(ctx, img.Name)
+	var imgID image.ID
+	if img == nil {
+		if len(all) == 0 {
+			return []imagetypes.DeleteResponse{}, nil
+		}
+		imgID = image.ID(all[0].Target.Digest)
+		var named reference.Named
+		if imageRef != string(imgID) {
+			parsedRef, err := reference.ParseNormalizedNamed(imageRef)
+			if err == nil {
+				named = parsedRef
+			}
+		}
+		sameRef, err := i.getSameReferences(ctx, named, all)
 		if err != nil {
 			return nil, err
 		}
-		i.LogImageEvent(imgID.String(), imgID.String(), events.ActionUnTag)
-		records := []imagetypes.DeleteResponse{{Untagged: reference.FamiliarString(reference.TagNameOnly(parsedRef))}}
-		return records, nil
-	}
 
-	using := func(c *container.Container) bool {
-		return c.ImageID == imgID
-	}
-	ctr := i.containers.First(using)
-	if ctr != nil {
-		if !force {
-			// If we removed the repository reference then
-			// this image would remain "dangling" and since
-			// we really want to avoid that the client must
-			// explicitly force its removal.
-			refString := reference.FamiliarString(reference.TagNameOnly(parsedRef))
-			err := &imageDeleteConflict{
-				reference: refString,
-				used:      true,
-				message: fmt.Sprintf("container %s is using its referenced image %s",
-					stringid.TruncateID(ctr.ID),
-					stringid.TruncateID(imgID.String())),
+		if len(sameRef) == 0 && named != nil {
+			return nil, dimages.ErrImageDoesNotExist{Ref: named}
+		}
+
+		if len(sameRef) == len(all) && !force {
+			c &= ^conflictActiveReference
+		}
+		if named != nil && len(sameRef) > 0 && len(sameRef) != len(all) {
+			familiarRef := reference.FamiliarString(named)
+			for _, ref := range sameRef {
+				// TODO: Add with target
+				err := i.images.Delete(ctx, ref.Name)
+				if err != nil {
+					return nil, err
+				}
 			}
+			i.logImageEvent(sameRef[0], familiarRef, events.ActionUnTag)
```

Comment:
Shouldn't the event be logged for every ref from `sameRef`?

Reply:
Probably, I'll update and see if any of the tests complain. Either I was worried that the log output would differ here or wasn't quite sure that it would matter.